ac_dc/config.py

```diff
@@ -0,0 +1,89 @@
+from logging import getLogger
+
+
+logger = getLogger(__name__)
+
+
+class BaseConfig:
+    """Base class for all pipeline configurations
+
+    Class attributes (overridden by derived classes)
+
+        - language_code: Two character language code (ISO 3166) for the pipeline
+
+    Parameters for filters:
+
+        - strip_characters: "' 0123456789¯_%$§½¼¾×|†~\"—±′–'°−{}[]·'?,./<>!@#^&*()+-‑=:;，：,...�`→¶.…‘’”"
+
+        - cond_check_empty: True
+
+        - cond_check_special_characters: True
+          - special_characters: "' 0123456789¯_%$§½¼¾×|†~\"—±′–'°−{}[]·'?,./<>!@#^&*()+-‑=:;，：,...�`→¶.…‘’”"
+          - special_characters_cutoff: 0.4
+
+        - cond_check_stopwords: True
+          - stopwords
+          - stopwords_cutoff: 0.4
+
+        - cond_check_badwords: True
+          - badwords
+          - badwords_cutoff: 0.4
+
+        - cond_check_lang_id: True
+          - path_model_fasttext: "/tmp/lid.176.bin"
+          - lang_id_cutoff: 0.8
+
+
+    Parameters for normalizers:
+
+        - cond_remove_words_with_incorrect_substrings": True,
+        - incorrect_word_substrings": ["http", "www", ".com", "href", "//"],
+
+        - cond_remove_long_words": False,
+        - length_word_cutoff": 50,
+
+    """
+
+    attribute_map = {}
+
+    def __setattr__(self, key, value):
+        if key in super().__getattribute__("attribute_map"):
+            key = super().__getattribute__("attribute_map")[key]
+        super().__setattr__(key, value)
+
+    def __getattribute__(self, key):
+        if key != "attribute_map" and key in super().__getattribute__("attribute_map"):
+            key = super().__getattribute__("attribute_map")[key]
+        return super().__getattribute__(key)
+
+    def __init__(self, language_code, **kwargs):
+        self.language_code = language_code
```

Comment:
can we make this a dataclass instead?


```python
@dataclass
class BaseConfig:
   language_code: str
   strip_character: bool
   ...
```

Reply:
Yeah that's right! dataclass will be more readable

Reply:
It's nice to use `datasets.map` for normalize and `datasets.filter` for filter, we can also run it in parallel. 

For the deduplication we should figure out where it would be in the pipeline. 

For example in [CCNet](https://arxiv.org/pdf/1911.00359.pdf), they perform deduplication first before lang_id and perplexity calculation (I'm not sure normalize before or after deduplication).